File: python-service/app/models/address.py
"""Pydantic models for address data structures."""

from typing import Optional

from pydantic import BaseModel


class StructuredAddress(BaseModel):
    """A structured US address with discrete fields."""

    street_line_1: str
    street_line_2: Optional[str] = None
    city: str
    state: str
    zipcode: str

    def to_string(self) -> str:
        """Serialize the address to a pipe-delimited string for round-trip support.

        Format: "street_line_1|street_line_2|city|state|zipcode"
        When street_line_2 is None, it is serialized as an empty string.
        """
        line_2 = self.street_line_2 if self.street_line_2 is not None else ""
        return f"{self.street_line_1}|{line_2}|{self.city}|{self.state}|{self.zipcode}"

    @classmethod
    def from_string(cls, value: str) -> "StructuredAddress":
        """Parse a pipe-delimited string back into a StructuredAddress.

        Expected format: "street_line_1|street_line_2|city|state|zipcode"
        An empty street_line_2 segment is interpreted as None.

        Raises:
            ValueError: If the string does not contain exactly 5 pipe-delimited segments.
        """
        parts = value.split("|")
        if len(parts) != 5:
            raise ValueError(
                f"Expected 5 pipe-delimited segments, got {len(parts)}: {value!r}"
            )
        street_line_1, street_line_2_raw, city, state, zipcode = parts
        street_line_2 = street_line_2_raw if street_line_2_raw != "" else None
        return cls(
            street_line_1=street_line_1,
            street_line_2=street_line_2,
            city=city,
            state=state,
            zipcode=zipcode,
        )

File: express-service/src/models/address.ts
/**
 * Structured address model for USPS address validation.
 */
export interface StructuredAddress {
  street_line_1: string;
  street_line_2?: string;
  city: string;
  state: string;
  zipcode: string;
}

const DELIMITER = '|';

/**
 * Serialize a StructuredAddress to a pipe-delimited string.
 * Format: "street_line_1|street_line_2|city|state|zipcode"
 * If street_line_2 is undefined or empty, it is serialized as an empty segment.
 */
export function serializeAddress(address: StructuredAddress): string {
  const streetLine2 = address.street_line_2 ?? '';
  return [
    address.street_line_1,
    streetLine2,
    address.city,
    address.state,
    address.zipcode,
  ].join(DELIMITER);
}

/**
 * Parse a pipe-delimited string back into a StructuredAddress.
 * Expects exactly 5 segments: street_line_1|street_line_2|city|state|zipcode
 * Throws an error if the format is invalid.
 */
export function parseAddress(str: string): StructuredAddress {
  const parts = str.split(DELIMITER);
  if (parts.length !== 5) {
    throw new Error(
      `Invalid address string: expected 5 pipe-delimited segments, got ${parts.length}`
    );
  }

  const [street_line_1, street_line_2, city, state, zipcode] = parts;

  return {
    street_line_1,
    ...(street_line_2 ? { street_line_2 } : {}),
    city,
    state,
    zipcode,
  };
}
